rdata/conversion/to_r.py

```diff
@@ -24,18 +27,31 @@
 )
 
 if TYPE_CHECKING:
-    from collections.abc import Mapping
+    from collections.abc import Callable, Mapping
     from typing import Any, Final, Literal, Protocol
 
-    Encoding = Literal["utf-8", "cp1252"]
+    import numpy.typing as npt
+
+    from rdata.unparser import FileType
 
+    Encoding = Literal["utf-8", "cp1252"]
 
     class Converter(Protocol):
-        """Protocol for Py-to-R conversion."""
+        """Protocol for class converting Python objects to R objects."""
+
+        format_version: int
+
+        def convert_to_r_sym(self, name: str) -> RObject:
+            """Convert string to R symbol."""
 
-        def __call__(self, data: Any, *, encoding: Encoding) -> RObject: # noqa: ANN401
-            """Convert Python object to R object."""
+        def convert_to_r_object(self, data: Any) -> RObject:  # noqa: ANN401
+            """Convert Python data to R object."""
 
+    ConstructorReturnValue = tuple[RObjectType, Any, dict[str, Any] | None]
+    ConstructorFunction1 = Callable[[Any], ConstructorReturnValue]
+    ConstructorFunction2 = Callable[[Any, Converter], ConstructorReturnValue]
```

Comment:
For writing, as this is a new API, I think we can force the signature to always include the `Converter`. For reading we can't do that, as there are existing constructors without that parameter, and thus we would have to add the new signature and deprecate the old one.